src/Oauth2Client.php

```diff
@@ -0,0 +1,320 @@
+    $redirectUrl,
+    $accessToken = null,
+    $refreshToken = null
+  ) {
+    $this->clientId = $clientId;
+    $this->clientSecret = $clientSecret;
+    $this->redirectUrl = $redirectUrl;
+    $this->accessToken = $accessToken;
+    $this->refreshToken = $refreshToken;
+  }
+
+  /**
+   * Grant Type 1: Resource owner credentials grant
+   *
+   * https://developers.shapeways.com/quick-start#authenticate
+   *
+   * Your don't need a redirect url for this grant type
+   *
+   * Use "access_token" from result for other API calls
+   *
+   * @return array - json decoded api response
+   */
+  public function generateAccessTokenClientCredentialGrant()
+  {
+    $params = array(
+      "grant_type" => "client_credentials"
+    );
+
+    $url = $this->baseApiUrl . '/oauth2/token';
+
+
+    $process = curl_init($url);
+    curl_setopt($process, CURLOPT_USERPWD, $this->clientId . ":" . $this->clientSecret);
+    curl_setopt($process, CURLOPT_TIMEOUT, 30);
+    curl_setopt($process, CURLOPT_POST, 1);
+    curl_setopt($process, CURLOPT_POSTFIELDS, $params);
+    curl_setopt($process, CURLOPT_RETURNTRANSFER, true);
+    $result = curl_exec($process);
+    curl_close($process);
+
+    return json_decode($result);
+
+  }
+
+
+  /**
+   * Grant Type 2: Authorization code grant
+   * Step 1 - generate a authorization code
+   *
+   *
+   * If you want to use this app for a different Shapeways User account.
+   *
+   *
+   * @link https://developers.shapeways.come/quick-start#authenticate
+   *
+   * redirect_uri - This determines where authentication requests will be sent and received by your app.
+   * Users will be redirected to this URL when they attempt to use your app.
+   */
+  public function generateAccessTokenAuthorizationGrant()
+  {
+    $clientId = $this->clientId;
+    $url = $this->baseApiUrl . "/oauth2/authorize?response_type=code&client_id=" . $clientId . "&redirect_uri=" . rawurlencode($this->redirectUrl);
+    header('Location: ' . $url);
+  }
+
+
+  /**
+   * Grant Type 2: Authorization code grant
+   * Step 2 - Generate access token
+   *
+   * @link https://developers.shapeways.com/quick-start#authenticate
+   *  function for handling call back of generateAccessTokenAuthorizationGrant() call above
+   *
+   * Use "access_token" from result for other API calls
+   *
+   * @return array - json decoded api response
+   */
+  public function handleAuthorizationGrantCallback()
+  {
+    $code = $_REQUEST['code'] ?? null;
+    $redirectURL = $this->redirectUrl;
+
+    if ($code === null) {
+      echo "Missing authorization code";
+      exit();
+    }
+    $url = $this->baseApiUrl . '/oauth2/token';
+
+    $data = array(
+      "grant_type" => "authorization_code",
+      "code" => $code,
+      "client_id" => $this->clientId,
+      "client_secret" => $this->clientSecret,
+      "redirect_uri" => $redirectURL
+    );
+
+    $ch = curl_init($url);
+    curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
+    curl_setopt($ch, CURLOPT_POST, 1);
```

Comment:
this is partially duplicate code with lines 95-99. Extract initiating the curl into its own function so the settings are consistent.